### PR TITLE
Better handling of uninitialized TlsContexts

### PR DESCRIPTION
**Pull request by rccarper**
*Issue #, if available:*
https://github.com/awslabs/aws-crt-cpp/issues/177

*Description of changes:*
* Calling NewConnectionOptions on a TlsContext that's invalid (null context and/or initialization error) will now result in an error message and TlsConnectionOptions that are "uninitialized", ie, created with the no-arg constructor.
  * Previously, this was always assuming the context was non-null, which can cause a crash.
  * I went back-and-forth on using an assert instead of the above.  I don't necessarily disagree with that; I do like how that would be louder.  I ultimately went this way because I think it's awkward for us to make the no-arg constructor available (which sets it as "uninitialized" and with a null context) and guarantee a crash when someone calls NewConnectionOptions.  (Although we do certainly have similar behavior on TlsContextOptions.)
* Passing in an invalid TlsContext to MqttClient::NewConnection will now result in a null connection being returned.
  * I think that this makes sense, given that we provide an overload for NewConnection that does not require a TlsContext.
  * I was initially concerned that this might technically introduce a new crash for people currently passing in invalid TlsContexts, but those people should already be crashing when this flow eventually calls NewConnectionOptions.
  * I did look at making similar changes to the HTTP Connection/Manager code.  However, those take in an already built TlsConnectionOptions, which are allowed to be uninitialized (ie: uninitialized=don't use TLS).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/io/TlsOptions.cpp [hunk 1]
```diff
@@ -283,6 +284,13 @@ namespace Aws
 
             TlsConnectionOptions TlsContext::NewConnectionOptions() const noexcept
             {
+                if (!*this)
```
Comment (justinboswell) [inline source/io/TlsOptions.cpp:287]: This gives me...feelings, but I get why it's here. Syntactically, it's a bit... obtuse. Is there not an IsValid() or similar we could call instead?

Comment (rccarper) [reply source/io/TlsOptions.cpp:287]: Negative, the class just overloads the operator right now.  I can certainly add one though!

**Review by justinboswell — APPROVED** [review]

**Review by graebm — COMMENTED** [review]
I'm not against this change, it's just that every function in this library has this issue 😕.
- Every member function could check that it's running on a valid instance.
- Every function that takes arguments could check that they are valid instances.

We chose not to use exceptions, so this is the world we live in.

File: source/mqtt/MqttClient.cpp [hunk 2]
```diff
@@ -694,6 +694,15 @@ namespace Aws
                 const Crt::Io::TlsContext &tlsContext,
                 bool useWebsocket) noexcept
             {
+                if (!tlsContext)
+                {
+                    AWS_LOGF_ERROR(
+                        AWS_LS_MQTT_CLIENT,
+                        "id=%p Trying to call MqttClient::NewConnection using an invalid TlsContext.",
+                        (void *)m_client);
+                    return nullptr;
```
Comment (graebm) [inline source/mqtt/MqttClient.cpp:703]: you also need to `aws_raise_error(AWS_ERROR_INVALID_ARGUMENT)`

Comment (rccarper) [reply source/mqtt/MqttClient.cpp:704]: Added!

**Comment by rccarper** [comment]
> I'm not against this change, it's just that every function in this library has this issue 😕.
> 
> * Every member function could check that it's running on a valid instance.
> * Every function that takes arguments could check that they are valid instances.
> 
> We chose not to use exceptions, so this is the world we live in.

I see where you're coming from.  I think minimally, MqttClient::NewConnection should still be validating its inputs.  I feel less strongly about the check in NewConnectionOptions, if we all feel that it's inconsistent/unnecessary.  But I do think we should at least be asserting in these cases.

**Review by graebm — APPROVED** [review]

**Pull request merged**